### PR TITLE
deprecate support to Envoy SDS v2 API

### LINKED ISSUE #4204: Deprecate SPIRE Agent Envoy SDS v2 API
In v1.7.0, when the v2 SDS API is invoked, log and return an error mentioning that the v2 API is deprecated, with a link pointing the user to migrating to v3 SDS. Introduce a configuration field to give the user a way to re-enable the v2 API, if needed.

### PR DESCRIPTION
**Pull request by MarcosDY**
Deprecate SDS v2 API, disabling it as default, but add a configurable to reactivate it.

**Which issue this PR fixes**
fixes #4204

### CODE REVIEW

File: pkg/agent/endpoints/sdsv2/handler.go [hunk 1]
```diff
@@ -204,6 +213,10 @@ func (h *Handler) DeltaSecrets(discovery_v2.SecretDiscoveryService_DeltaSecretsS
 }
```
Comment (rturner3) [inline pkg/agent/endpoints/sdsv2/handler.go:213]: Shall we also return the same `Unavailable` code for the `DeltaSecrets` method when the SDS v2 API is not enabled to be consistent across all the methods in the SDS v2 API?

File: pkg/agent/endpoints/sdsv2/handler_test.go [hunk 2]
```diff
@@ -204,6 +206,24 @@ func (s *HandlerSuite) TestStreamSecretsStreamAllSecrets() {
 	s.requireSecrets(resp, tdValidationContext, fedValidationContext, workloadTLSCertificate1)
 }
 
+func (s *HandlerSuite) TestAPIIsNoEnabled() {
```
Comment (rturner3) [inline pkg/agent/endpoints/sdsv2/handler_test.go:209]:
```suggestion
func (s *HandlerSuite) TestAPIIsNotEnabled() {
```

File: test/integration/suites/envoy-sds-v2/conf/downstream-agent/agent.conf [hunk 3]
```diff
@@ -6,6 +6,9 @@ agent {
     socket_path ="/opt/shared/agent.sock"
     trust_bundle_path = "/opt/spire/conf/agent/bootstrap.crt"
     trust_domain = "domain.test"
+    sds = {
+        enable_sds_v2 = true
```
Comment (rturner3) [inline test/integration/suites/envoy-sds-v2/conf/downstream-agent/agent.conf:10]:
```suggestion
        enable_deprecated_v2_api = true
```

File: test/integration/suites/envoy-sds-v2/conf/upstream-agent/agent.conf [hunk 4]
```diff
@@ -6,6 +6,9 @@ agent {
     socket_path ="/opt/shared/agent.sock"
     trust_bundle_path = "/opt/spire/conf/agent/bootstrap.crt"
     trust_domain = "domain.test"
+    sds = {
+        enable_sds_v2 = true
```
Comment (rturner3) [inline test/integration/suites/envoy-sds-v2/conf/upstream-agent/agent.conf:10]:
```suggestion
        enable_deprecated_sds_v2_api = true
```

File: doc/spire_agent.md [hunk 5]
```diff
@@ -84,12 +84,13 @@ Only one of these three options may be set at a time.
...
-| `default_svid_name`              | The TLS Certificate resource name to use for the default X509-SVID with Envoy SDS                | default |
-| `default_bundle_name`            | The Validation Context resource name to use for the default X.509 bundle with Envoy SDS          | ROOTCA  |
-| `default_all_bundles_name`       | The Validation Context resource name to use for all bundles (including federated) with Envoy SDS | ALL     |
-| `disable_spiffe_cert_validation` | Disable Envoy SDS custom validation                                                              | false   |
+| Configuration                    | Description                                                                                                                         | Default |
+|----------------------------------|-------------------------------------------------------------------------------------------------------------------------------------|---------|
+| `enable_sds_v2`                  | Enable SDS v2 API. Please note that enabling SDS v2 is not recommended and this config setting will be removed in a future version. | false   |
```
Comment (rturner3) [inline doc/spire_agent.md:89]:
```suggestion
| `enable_deprecated_v2_api`                  | Enable deprecated SDS v2 API. It is recommended that users of the SDS v2 API migrate to the SDS v3 API. The SDS v2 API and this config setting will be removed in a future version. | false   |
```

File: cmd/spire-agent/cli/run/run_test.go [hunk 6]
```diff
@@ -244,6 +244,18 @@ func TestMergeInput(t *testing.T) {
 				require.Equal(t, "foo", c.Agent.SDS.DefaultAllBundlesName)
 			},
 		},
+		{
+			msg: "enable_sds_v2 should be configurable by file",
```
Comment (rturner3) [inline cmd/spire-agent/cli/run/run_test.go:248]:
```suggestion
			msg: "enable_deprecated_v2_api should be configurable by file",
```

File: conf/agent/agent_full.conf [hunk 7]
```diff
@@ -72,6 +72,10 @@ agent {
 
     # sds: Optional SDS configuration section.
     # sds = {
+    #     # enable_sds_v2: Enable SDS v2 API. Please note that enabling SDS v2 is not 
+    #     # recommended and this config setting will be removed in a future version. Default: false
+    #     # enable_sds_v2 = false
```
Comment (rturner3) [inline conf/agent/agent_full.conf:77]:
```suggestion
    #     # enable_deprecated_v2_api = false
```

File: conf/agent/agent_full.conf [hunk 8]
```diff
@@ -72,6 +72,10 @@ agent {
 
     # sds: Optional SDS configuration section.
     # sds = {
+    #     # enable_sds_v2: Enable SDS v2 API. Please note that enabling SDS v2 is not 
+    #     # recommended and this config setting will be removed in a future version. Default: false
```
Comment (rturner3) [inline conf/agent/agent_full.conf:76]:
```suggestion
    #     # enable_deprecated_v2_api: Enable deprecated SDS v2 API. It is recommended that
    #     # users of the SDS v2 API migrate to the SDS v3 API.
    #     # The SDS v2 API and this config setting will be removed in a future version.
    #     # Default: false
```

File: pkg/agent/endpoints/sdsv2/handler.go [hunk 9]
```diff
@@ -26,6 +26,10 @@ import (
 	"google.golang.org/protobuf/types/known/anypb"
 )
 
+const (
+	deprecatedAPIErrorMsg = "support to Envoy SDS v2 API has been deprecated in SPIRE and is no longer supported by Envoy. Please refer to: https://www.envoyproxy.io/docs/envoy/latest/api/api_supported_versions. It can be enabled using the config setting 'enable_sds_v2': https://github.com/spiffe/spire/blob/main/doc/spire_agent.md#sds-configuration. Please note that enabling SDS v2 is not recommended and this config setting will be removed in a future version."
```
[the remark this answers is not in the document]
Comment (MarcosDY) [reply pkg/agent/endpoints/sdsv2/handler.go:30]: it is an error, so I'll change the message  start from `The` to `the`

**Pull request merged**